playwright/_impl/_browser.py

```diff
@@ -148,6 +148,12 @@ async def new_page(
     async def close(self) -> None:
         if self._is_closed_or_closing:
             return
+        if self._is_remote:
```

Comment:
It's possible that the WebSocket connection get's disconnected without the user's control. This will end up that the Browser.close method get's never called. Moving the loop to when the WebSocket connection actually closes would fix it, see here how it's done upstream:

https://github.com/microsoft/playwright/blob/e018c5dd426c7cef9bbf6eb497112083013bb6ba/src/client/browserType.ts#L170-L179

Seems like there is no test yet for it, having one for that would be neat.

Reply:
Upstream we have this loop only a single time, would be good if we could have it here also at a single place only. It feels wrong to have it inside: _on_close.

Reply:
When the transport closes it calls `_on_close` when websocket connection is closed

Reply:
Upstream we do it next to the ws close event and not in the browser, I would move it into our browser_type.py when we receive the disconnect event from the WS client.

See here how it's done upstream: https://github.com/microsoft/playwright/blob/4c3bd11820665274071ee24663a58594a0b34b94/src/client/browserType.ts#L168-L177

Reply:
It uses event listener for websocket close but in python we don't have that, for that a custom protocol class would be required which on close would do that

Reply:
We have it, see here :) https://github.com/microsoft/playwright-python/blob/master/playwright/_impl/_browser_type.py#L212

Reply:
In my opinion we still should move this code away from the browser and into connect method as it was done upstream since it's specific to remote connections via connect.